### DISCUSSION TITLE
Can't upload the image

### DISCUSSION THREAD

**Discussion by DandeMC** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#87] [Q&A]
Hello,
I followed the instructions in the pdf file, but I can't upload the image. 

```
`Building in release mode
Retrieving maximum program size .pio\build\nodemcuv2\firmware.elf
Checking size .pio\build\nodemcuv2\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=====     ]  48.3% (used 39572 bytes from 81920 bytes)
Flash: [====      ]  40.9% (used 426904 bytes from 1044464 bytes)
Configuring upload protocol...
AVAILABLE: espota, esptool
CURRENT: upload_protocol = espota
Uploading .pio\build\nodemcuv2\firmware.bin
08:56:26 [DEBUG]: Options: {'esp_ip': '192.168.4.121', 'host_ip': '0.0.0.0', 'esp_port': 8266, 'host_port': 37814, 'auth': 'esp8266', 'image': '.pio\\build\\nodemcuv2\\firmware.bin', 'spiffs': False, 'debug': True, 'progress': True}
08:56:26 [INFO]: Starting on 0.0.0.0:37814
08:56:26 [INFO]: Upload size: 431056
08:56:26 [INFO]: Sending invitation to: 192.168.4.121
08:56:36 [ERROR]: No Answer
*** [upload] Error 1
================================================================================================================================ [FAILED] Took 11.46 seconds ================================================================================================================================The terminal process "C:\Users\ml\.platformio\penv\Scripts\platformio.exe 'run', '--target', 'upload'" terminated with exit code: 1.

Terminal will be reused by tasks, press any key to close it.`
```

What can I do?

**Comment by visualapproach** [comment]
08:56:26 [INFO]: Sending invitation to: 192.168.4.121
08:56:36 [ERROR]: No Answer

The platformio.ini file is telling your computer to upload image over the air to that particurlar IP address. Strangely it's the same address I had on my device when I last saved... 🤔 
Just comment out the OTA lines in the file and use USB the first time you upload. (Or download the new file from github development branch) The following uploads can be made over the air if you set the right IP.
The file should look like this:
.
.
.
;uncomment and edit IP for OTA upload
; upload_protocol = espota
; upload_port = 192.168.4.121
; upload_flags = 
; 	--auth=esp8266

**Comment by DandeMC** [comment]
wow. thx for your fast replay. now I have the next error... I think I must the usb port in visual studio? but where?

```
Looking for upload port...
Error: Please specify `upload_port` for environment or use global `--upload-port` option.
For some development platforms it can be a USB flash drive (i.e. /media/<user>/<device name>)
*** [upload] Explicit exit, status 1
```

**Reply by visualapproach** [reply]
It should not be needed since it automatically scans the COM ports. You can specify it though if you know what port your device is connected to.
For example:
`upload_port = COM10`

Did you save the platformio.ini file?
Maybe restart VSC and retry.
In some devices (you haven't told what device you are using) you must press reset button just before upload.

**Reply by DandeMC** [reply]
I bought this hiere:
https://www.amazon.de/gp/product/B06Y1ZPNMS/ref=ppx_yo_dt_b_asin_title_o02_s00?ie=UTF8&psc=1

where can I specify upload_port? can it be that i need drivers for the device?

**Reply by visualapproach** [reply]
Ok,
Where to put upload_port parameter; platformio.ini

Need drivers; maybe but I think it's part of windows. You could check if your chip has a CP2101 (or whatever it's called) or a ch340 and google for drivers. But I think you should begin with checking your USB cable. Many cheap ones only has power and gnd to charge things, not data wires. When you know the cable is working (but still no go in VSC) you could try uploading something from arduino ide to see if it works.

**Reply by DandeMC** [reply]
Ok, it was the usb cable aaand the driver... now I can see it as COM3, but the next error came

```
Serial port COM3
Traceback (most recent call last):
  File "C:\Users\ml\.platformio\packages\tool-esptoolpy\esptool.py", line 3969, in <module>
    _main()
  File "C:\Users\ml\.platformio\packages\tool-esptoolpy\esptool.py", line 3962, in _main
    main()
  File "C:\Users\ml\.platformio\packages\tool-esptoolpy\esptool.py", line 3551, in main
    esp = chip_class(each_port, initial_baud, args.trace)
  File "C:\Users\ml\.platformio\packages\tool-esptoolpy\esptool.py", line 271, in __init__
    self._port = serial.serial_for_url(port)
  File "c:\users\ml\.platformio\penv\lib\site-packages\serial\__init__.py", line 90, in serial_for_url
    instance.open()
  File "c:\users\ml\.platformio\penv\lib\site-packages\serial\serialwin32.py", line 80, in open
    self._reconfigure_port()
  File "c:\users\ml\.platformio\penv\lib\site-packages\serial\serialwin32.py", line 222, in _reconfigure_port
    raise SerialException(
serial.serialutil.SerialException: Cannot configure port, something went wrong. Original message: PermissionError(13, 'Ein an das System angeschlossenes Gerät funktioniert nicht.', None, 31)
*** [uploadfs] Error 1
================================================================================================================================ [FAILED] Took 2.44 seconds ================================================================================================================================
```

**Reply by DandeMC** [reply]
never mind. now it works. PC restart helps every time. THX for your support =)
